api/src/blocks.rs

```diff
@@ -0,0 +1,167 @@
+}
+
+async fn handle_get_block(
+    ledger_version: Option<LedgerVersionParam>,
+    context: Context,
+) -> Result<impl Reply, Rejection> {
+    fail_point("endpoint_get_block")?;
+    Ok(Block::new(ledger_version, context)?.account()?)
+}
+
+async fn handle_get_genesis_ledger_info(context: Context) -> Result<impl Reply, Rejection> {
+    fail_point("endpoint_get_genesis_ledger_info")?;
+    Ok(context.get_genesis_ledger_info()?)
+}
+
+pub struct BlockDescription {
+    starting_version: u64,
+    ending_version: u64,
+    num_transactions: u64,
+    hash: HashValue,
+}
+
+pub(crate) struct Block {
+    ledger_version: u64,
+    latest_ledger_info: LedgerInfo,
+    context: Context,
+}
+
+impl Block {
+    pub fn new(
+        ledger_version: Option<LedgerVersionParam>,
+        context: Context,
+    ) -> Result<Self, Error> {
+        let latest_ledger_info = context.get_latest_ledger_info()?;
+        let ledger_version = ledger_version
+            .map(|v| v.parse("ledger version"))
+            .unwrap_or_else(|| Ok(latest_ledger_info.version()))?;
+
+        if ledger_version > latest_ledger_info.version() {
+            return Err(Error::not_found(
+                "ledger",
+                TransactionId::Version(ledger_version),
+                latest_ledger_info.version(),
+            ));
+        }
+
+        Ok(Self {
+            ledger_version,
+            latest_ledger_info,
+            context,
+        })
+    }
+
+    /// Scans the DB
+    pub fn find_block(self) -> Result<BlockDescription, Error> {
+        // Genesis is as special case, always version 0
+        if self.ledger_version == 0 {
+            let hash = self.context.get_genesis_accumulator_hash()?;
+            return Ok(BlockDescription {
+                starting_version: 0,
+                ending_version: 0,
+                num_transactions: 1,
+                hash: hash.into(),
+            });
+        }
+
+        // Every other block needs to be found
+        const MAX_BLOCK_SIZE: u16 = 10000;
+        let search_start = self.ledger_version.saturating_sub(MAX_BLOCK_SIZE as u64);
+        let txns =
+            self.context
+                .get_transactions(search_start, MAX_BLOCK_SIZE * 2, self.ledger_version)?;
+
+        // Search for the closest block boundaries to the version
+        let mut start = search_start;
+        let mut end = search_start.saturating_add((MAX_BLOCK_SIZE * 2) as u64);
+        let mut hash = None;
+        for txn in txns {
+            let txn_version = txn.version;
+            match &txn.transaction {
+                Transaction::BlockMetadata(_) => {
+                    // BlockMetadata is the beginning of the block
+                    if start < txn_version && txn_version <= self.ledger_version {
+                        start = txn_version;
+                    }
+                }
+                Transaction::StateCheckpoint => {
+                    // StateCheckpoint is at the end of the block
```

Comment:
no, the state checkpoint is not a block epilog, don't assume it is a block boundary. We are ready to stop issuing state checkpoints as a way to boost performance in emergency.